### DISCUSSION TITLE
Build and launch speed

### DISCUSSION THREAD

**Discussion by nref** [AvaloniaUI/Avalonia#11046] [Q&A] [answered]
Hi, and thanks for all the work on the browser in Avalonia 11.

I've got Visual Studio 2022 17.5.4, AvaloniaVersion `11.0.0-preview5`, `net7.0`, and a very capable machine (32 real cores, 128GB RAM). 

I've noticed that building the browser target is pretty slow. 

Repro:

Create an Avalonia project with `dotnet new avalonia.xplat`. Select Debug configuration. Build the generated Browser project. Notice it takes about 10-15 seconds. 

This combined with another 5-10s launch time makes the edit-compile-debug cycle pretty slow. It's hard to get into a flow. Release takes much longer, 30-40 seconds, which is acceptable for a build pipeline. I'm only really talking about Debug.

What are the hopes for improving the build or launch speed?

I noticed this line in the build log

`Compiling native assets with emcc with -O0. This may take a while ...`

Are there any tricks to employ for speeding up the emscripten compiler frontend?

I've got the Avalonia VS extension, but to really test my app e.g. with web requests, local storage, etc, I have to launch it.

For the launch speed, what about live / hot reloading?

**Comment by maxkatz6** [comment, marked as answer]
Hello!
Avalonia VS extension is unlikely to slow it down that much.
You should run your debug build with binary logger tool, see https://msbuildlog.com/. It will show how much time is spend on which task.

There is a chance, that the build is slowed down by XAML compilation, but it's unlikely for the small xplat project.
Also, if it's the case, normal desktop app build would also be slow.

Most likely, this issue cannot be solved by Avalonia team, as we don't maintain "edit-compile-debug" cycle for the .NET on WASM.
You can fill an issue in official runtime repository https://github.com/dotnet/runtime/issues?q=is%3Aissue+is%3Aopen+label%3Aarch-wasm+

**Reply by maxkatz6** [reply]
Out of curiosity, checked what takes lots of time on my local machine. First build on i9-13900 took 20 seconds. Second build was only 5 seconds (still not perfect, but way better).

For the first build, I see following statistic using MSBuildStructuredLogViewer:
![image](https://user-images.githubusercontent.com/3163374/232958772-8451e49f-527a-4b02-952b-c0a98b877d2c.png)

You can see that C# compiler (Csc) and Avalonia XAML compiler are quite fast, and most of the time was spent on WASM related compilation. In Debug mode.

**Comment by nref** [comment]
Thanks Max for the quick reply. And thanks for teaching me about msbuildlog. As I suspected, the WASM compiler (emscripten) is what takes so long. I guess I just have to accept that as the current state of WASM.

For workarounds, the the Avalonia VS extension reduces marginally how often I have to build and launch. But is hot reloading on the roadmap? Blazor and Rust+Yew+Trunk have it, but I really prefer XAML!

Thanks for all you do on Avalonia, this is a fantastic project.

**Reply by Tamnac** [reply]
There's an experimental hot reload here https://github.com/AvaloniaUI/Live.Avalonia

**Comment by robloo** [comment]
I would suggest developing on a desktop platform until the application is completed. Then switch over to WASM for debugging and finishing touches. Hopefully .NET itself gets improvements here someday but I expect they will need a new, unified compiler first which is going to be a long time (if ever).
